x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -191,51 +201,185 @@ public void authorize(final Authentication authentication, final String action,
-                } else {
-                    auditTrailService.get().tamperedRequest(null, authentication, action, originalRequest);
-                    final String message = "Attempt to authorize action [" + action + "] for [" + authentication.getUser().principal()
-                            + "] without an existing request-id";
-                    assert false : message;
-                    listener.onFailure(new ElasticsearchSecurityException(message));
-                    return;
-                }
-            }
+    private void authorize(final Authentication authentication, final String action, final TransportRequest originalRequest,
+                           final ActionListener<Void> listener,
+                           final TriConsumer<RequestInfo, String, ActionListener<Void>> actionAuthorization) {
+        String auditId;
+        try {
+            auditId = requireAuditId(authentication, action, originalRequest);
+        } catch (ElasticsearchSecurityException e) {
+            listener.onFailure(e);
+            return;
+        }
+
+        // sometimes a request might be wrapped within another, which is the case for proxied
+        // requests and concrete shard requests
+        final TransportRequest unwrappedRequest = maybeUnwrapRequest(authentication, originalRequest, action, auditId);
 
-            // sometimes a request might be wrapped within another, which is the case for proxied
-            // requests and concrete shard requests
-            final TransportRequest unwrappedRequest = maybeUnwrapRequest(authentication, originalRequest, action, auditId);
-
-            // Check operator privileges
-            // TODO: audit?
-            final ElasticsearchSecurityException operatorException =
-                operatorPrivilegesService.check(action, originalRequest, threadContext);
-            if (operatorException != null) {
-                listener.onFailure(denialException(authentication, action, originalRequest, operatorException));
-                return;
+        if (checkOperatorPrivileges(authentication, action, originalRequest, listener) == false) {
+            return;
+        }
+
+        if (SystemUser.is(authentication.getUser())) {
+            // this never goes async so no need to wrap the listener
+            authorizeSystemUser(authentication, action, auditId, unwrappedRequest, listener);
+        } else {
+            final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action);
+            actionAuthorization.apply(requestInfo, auditId, listener);
+        }
+    }
+
+    /**
+     * @return {@code true} if this action is a child action of an already authorized action that is in the thread context
+     * <strong>and</strong> we can infer that this aciton is authorized from the parent action.
+     */
+    private boolean isPreauthorizedChildAction(Authentication authentication, String action, TransportRequest request) {
```

Comment:
It's probably better to make this method part of `AuthorizationEngine` and have the implementation provided by `RBACEngine`. There could be cases where a custom authz engine needs to behave differently.

Reply:
Yeah, fair enough. I was trying to keep it simple, but I think you're right.